### PR TITLE
[NET-2994] Change TCP Retransmit Counts

### PR DESCRIPTION
**Pull request by akarpz**
### What does this PR do?

[Read the RFC first](https://datadoghq.atlassian.net/wiki/spaces/NET/pages/2800026750/Counting+TCP+Retransmits+in+System+Probe+on+Linux#Approaches-Going-Forward)

This PR makes several changes to the way TCP retransmits are counted:

- **On Prebuilt:**
  - Adds a Kretprobe that checks the return code of `tcp_retransmit_skb`. If the error code is non-zero, we no longer increment the retransmit count by the value of `segs`. This excludes failed retransmits from the count.
- **On Runtime Compiled:** 
  - Also adds a Kretprobe, but uses the value `retrans_out` on the `struct sock` param instead of `segs`. `retrans_out` is only incremented when a retransmit was successfully sent. It's also a monotonic counter, so we calculate the delta between its value before `tcp_retransmit_skb` is executed and afterwards. 

In general these changes _reduce_ the number of retransmits we display in NPM.

**Other Notes:**

- It was observed during testing that the RC changes count SYN retransmits. This is different from the `system.net` metrics which exclude them. 
- This PR is similar but not identical to a previous change to TCP retransmit counts. That change was reverted due to concerns that we did not fully understand how we were changing the counts. After testing this change on staging we feel we can fully explain the changes.

There is more detail here: 

### Motivation

https://datadoghq.atlassian.net/browse/NET-2994

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by kevinconaway** [comment]
Would you please help me understand why we need two approaches, one for pre-built and one for CORE?

**Comment by akarpz** [comment]
@kevinconaway this is discussed a bit in the RFC, here is an excerpt:

> The kernel does keep counts of retransmitted segments on struct sock objects, namely retrans_out and total_retrans. The former is incremented when a retransmit is successfully queued, while the latter is incremented the same time and amount as RetransSegs. Note that these fields would not be available to the prebuilt variant of the tracer as they would require offset guessing while trying to set up scenarios to drop packets on customer machines.

Basically we only have access to the field we really want in RC, so we use a suboptimal approach in prebuilt. I believe this sort of thing is increasingly common as we phase out support for prebuilt.

File: releasenotes/notes/tcp-retransmit-updates-43d2b63ea00f750f.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - |
+    No longer increments tcp retransmit count when the retransmit fails.
```
Comment (buraizu) [inline releasenotes/notes/tcp-retransmit-updates-43d2b63ea00f750f.yaml:11]:
```suggestion
    No longer increments TCP retransmit count when the retransmit fails.
```

**Review by buraizu — APPROVED** [review]
Minor edit to the release note, but otherwise LGTM

**Review by hmahmood — DISMISSED** [review]

File: releasenotes/notes/tcp-retransmit-updates-43d2b63ea00f750f.yaml (hunk 1 above)
Comment (akarpz) [reply releasenotes/notes/tcp-retransmit-updates-43d2b63ea00f750f.yaml]: thanks, done

File: pkg/network/ebpf/c/co-re/tracer-fentry.c [hunk 2]
```diff
@@ -279,8 +279,40 @@ int BPF_PROG(udpv6_recvmsg_exit, struct sock *sk, struct msghdr *msg, size_t len
...
+    }
+
+    bpf_map_update_with_telemetry(pending_tcp_retransmit_skb, &tid, &args, BPF_ANY);
+
+    return 0;
+}
+
+SEC("fexit/tcp_retransmit_skb")
+int BPF_PROG(tcp_retransmit_skb_exit, struct sock *sk, struct sk_buff *skb, int segs, int err) {
+    log_debug("kretprobe/tcp_retransmit\n");
```
Comment (hmahmood) [inline pkg/network/ebpf/c/co-re/tracer-fentry.c:298]: kretprobe -> fexit

File: pkg/network/ebpf/c/co-re/tracer-fentry.c [hunk 3]
```diff
@@ -279,8 +279,40 @@ int BPF_PROG(udpv6_recvmsg_exit, struct sock *sk, struct msghdr *msg, size_t len
...
 SEC("fentry/tcp_retransmit_skb")
 int BPF_PROG(tcp_retransmit_skb, struct sock *sk, struct sk_buff *skb, int segs, int err) {
-    log_debug("fexit/tcp_retransmit: segs: %d\n", segs);
-    return handle_retransmit(sk, segs);
+    u64 tid = bpf_get_current_pid_tgid();
+    tcp_retransmit_skb_args_t args = {};
+    args.sk = sk;
+    args.segs = 0;
+    args.retrans_out_pre = 0;
+    if (bpf_probe_read_kernel_with_telemetry(&(args.retrans_out_pre), sizeof(args.retrans_out_pre), &(tcp_sk(sk)->retrans_out)) < 0) {
```
Comment (hmahmood) [inline pkg/network/ebpf/c/co-re/tracer-fentry.c:287]: Use `BPF_CORE_READ` or `BPF_CORE_READ_INTO` here.

File: pkg/network/ebpf/c/co-re/tracer-fentry.c [hunk 4]
```diff
@@ -279,8 +279,40 @@ int BPF_PROG(udpv6_recvmsg_exit, struct sock *sk, struct msghdr *msg, size_t len
 
 SEC("fentry/tcp_retransmit_skb")
 int BPF_PROG(tcp_retransmit_skb, struct sock *sk, struct sk_buff *skb, int segs, int err) {
-    log_debug("fexit/tcp_retransmit: segs: %d\n", segs);
-    return handle_retransmit(sk, segs);
+    u64 tid = bpf_get_current_pid_tgid();
+    tcp_retransmit_skb_args_t args = {};
+    args.sk = sk;
```
Comment (hmahmood) [inline pkg/network/ebpf/c/co-re/tracer-fentry.c:284]: No need to store anything other than `retrans_out_pre` here.

File: pkg/network/ebpf/c/co-re/tracer-fentry.c [hunk 5]
```diff
@@ -279,8 +279,40 @@ int BPF_PROG(udpv6_recvmsg_exit, struct sock *sk, struct msghdr *msg, size_t len
...
+        return 0;
+    }
+    tcp_retransmit_skb_args_t *args = bpf_map_lookup_elem(&pending_tcp_retransmit_skb, &tid);
+    if (args == NULL) {
+        return 0;
+    }
+    struct sock* sock = args->sk;
+    u32 retrans_out_pre = args->retrans_out_pre;
+    u32 retrans_out = 0;
+    bpf_probe_read_kernel_with_telemetry(&retrans_out, sizeof(retrans_out), &(tcp_sk(sock)->retrans_out));
```
Comment (hmahmood) [inline pkg/network/ebpf/c/co-re/tracer-fentry.c:311]: Use `BPF_CORE_READ` or `BPF_CORE_READ_INTO` here.

**Review by heyronhay — APPROVED** [review]
I read the code, I basically understood what was going on and it looks like typical ebpf code ;) That's about the extent of this review!

File: pkg/network/ebpf/c/co-re/tracer-fentry.c (hunk 3 above)
Comment (hmahmood) [reply pkg/network/ebpf/c/co-re/tracer-fentry.c:287]: Lots of details here: https://nakryiko.com/posts/bpf-core-reference-guide/#reading-kernel-data

tl;dr the macros make using `bpf_core_read` easier to use. `bpf_core_read` should be used here instead of `bpf_probe_read` since we want to record information about the field offsets so that they can be adjusted at runtime.

**Review by hmahmood — APPROVED** [review]
